x-pack/plugins/maps/public/classes/sources/mvt_single_layer_vector_source/mvt_field_config_editor.tsx

```diff
@@ -0,0 +1,225 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+/* eslint-disable @typescript-eslint/consistent-type-definitions */
+
+import React, { ChangeEvent, Component, Fragment } from 'react';
+import {
+  EuiButtonIcon,
+  EuiFlexGroup,
+  EuiFlexItem,
+  EuiButtonEmpty,
+  EuiSuperSelect,
+  EuiFieldText,
+  EuiSpacer,
+} from '@elastic/eui';
+import { i18n } from '@kbn/i18n';
+import _ from 'lodash';
+import { MVTFieldDescriptor } from '../../../../common/descriptor_types';
+import { FieldIcon } from '../../../../../../../src/plugins/kibana_react/public';
+import { MVT_FIELD_TYPE } from '../../../../common/constants';
+
+function makeOption({
+  value,
+  icon,
+  message,
+}: {
+  value: MVT_FIELD_TYPE;
+  icon: string;
+  message: string;
+}) {
+  return {
+    value,
+    inputDisplay: (
+      <EuiFlexGroup gutterSize="s" alignItems="center">
+        <EuiFlexItem grow={null}>
+          <FieldIcon type={icon} fill="none" />
+        </EuiFlexItem>
+        <EuiFlexItem>{message}</EuiFlexItem>
+      </EuiFlexGroup>
+    ),
+  };
+}
+
+const FIELD_TYPE_OPTIONS = [
+  {
+    value: MVT_FIELD_TYPE.STRING,
+    icon: 'string',
+    message: i18n.translate('xpack.maps.mvtSource.stringFieldLabel', {
+      defaultMessage: 'string',
+    }),
+  },
+  {
+    value: MVT_FIELD_TYPE.NUMBER,
+    icon: 'number',
+    message: i18n.translate('xpack.maps.mvtSource.numberFieldLabel', {
+      defaultMessage: 'number',
+    }),
+  },
+].map(makeOption);
+
+export interface Props {
+  fields: MVTFieldDescriptor[];
+  onChange: (fields: MVTFieldDescriptor[]) => void;
+}
+
+// eslint-disable-next-line @typescript-eslint/no-empty-interface
+interface State {
+  previousFields: MVTFieldDescriptor[];
+  currentFields: MVTFieldDescriptor[];
+}
+
+export class MVTFieldConfigEditor extends Component<Props, State> {
+  state: State = {
+    currentFields: [],
+    previousFields: [],
+  };
+
+  static getDerivedStateFromProps(nextProps: Props, prevState: State) {
+    if (_.isEqual(nextProps.fields, prevState.previousFields)) {
+      return null;
+    }
+    const clonedFields = _.cloneDeep(nextProps.fields);
+    return {
+      currentFields: clonedFields,
+      previousFields: clonedFields,
```

Comment:
We could probably just store `currentFields` in state. `previousFields` is really only used for naming purposes and isn't used outside of this function. I think it would be equally readable to check `if (_.isEqual(nextProps.fields, prevState.currentFields))` on line 81.